atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/LocalWriteBuffer.java

```diff
@@ -0,0 +1,118 @@
+        int numMetadataWritten = 0;
+        synchronized (getLockForTable(tableRef)) {
+            for (Map.Entry<Cell, byte[]> e : values.entrySet()) {
+                byte[] val = MoreObjects.firstNonNull(e.getValue(), PtBytes.EMPTY_BYTE_ARRAY);
+                Cell cell = e.getKey();
+                byte[] oldVal = writes.put(cell, val);
+                // If we are not writing metadata for a value, we have to remove any previously stored metadata since
+                // it may not be valid for the new value.
+                if (metadata.containsKey(cell)) {
+                    metadataForWrites.put(cell, metadata.get(cell));
+                    numMetadataWritten++;
+                } else {
+                    metadataForWrites.remove(cell);
+                }
+                long toAdd = val.length + Cells.getApproxSizeOfCell(cell);
+                long toSubtract = oldVal != null ? oldVal.length + Cells.getApproxSizeOfCell(cell) : 0;
+                long newByteCount = byteCount.addAndGet(toAdd - toSubtract);
+                if (newByteCount >= TransactionConstants.WARN_LEVEL_FOR_QUEUED_BYTES
+                        && newByteCount - toAdd < TransactionConstants.WARN_LEVEL_FOR_QUEUED_BYTES) {
+                    log.warn(
+                            "A single transaction has put quite a few bytes: {}. "
+                                    + "Enable debug logging for more information",
+                            SafeArg.of("numBytes", newByteCount));
+                    if (log.isDebugEnabled()) {
+                        log.debug(
+                                "This exception and stack trace are provided for debugging purposes.",
+                                new RuntimeException());
+                    }
+                }
+            }
+        }
+        if (numMetadataWritten != metadata.size()) {
+            Set<Cell> cellsWithOnlyMetadata = Sets.difference(metadata.keySet(), values.keySet());
+            throw new SafeIllegalStateException(
+                    "Every metadata we put must be associated with a write",
+                    UnsafeArg.of("tableRef", tableRef),
+                    UnsafeArg.of("cellsWithOnlyMetadata", cellsWithOnlyMetadata));
+        }
+    }
+
+    public ConcurrentMap<TableReference, ConcurrentNavigableMap<Cell, byte[]>> getLocalWrites() {
+        return writesByTable;
+    }
+
+    public ConcurrentNavigableMap<Cell, byte[]> getLocalWritesForTable(TableReference tableRef) {
+        return writesByTable.computeIfAbsent(tableRef, unused -> new ConcurrentSkipListMap<>());
+    }
+
+    // No need for concurrency control on the cell level since it is only written to with a lock and
+    // read during commit, which is guaranteed to be single-threaded and exclusive with writing.
+    public Map<Cell, ChangeMetadata> getChangeMetadataForWritesToTable(TableReference tableRef) {
+        return metadataByTable.computeIfAbsent(tableRef, unused -> new HashMap<>());
```

Comment:
Should we wrap these with `Collections.unmodifiableMap` ?

Reply:
(Not doing for local-writes, as it's used so widely, and is out of scope for this code review)